src/java/org/jivesoftware/openfire/plugin/util/cache/ClusterListener.java

```diff
@@ -151,21 +153,49 @@ public void memberAdded(final MembershipEvent event) {
         final NodeID nodeID = ClusteredCacheFactory.getNodeID(event.getMember());
         if (event.getMember().localMember()) { // We left and re-joined the cluster
             joinCluster();
+
+            waitForClusterCacheToBeInstalled();
+
+            // Let the other nodes know that we joined the cluster
+            logger.debug("Done joining the cluster. Now proceed informing other nodes that we joined the cluster.");
+            CacheFactory.doClusterTask(new NewClusterMemberJoinedTask(true));
         } else {
-            if(wasSenior && !isSenior) {
+            if (wasSenior && !isSenior) {
                 logger.warn("Recovering from split-brain; firing leftCluster()/joinedCluster() events");
                 ClusteredCacheFactory.fireLeftClusterAndWaitToComplete(Duration.ofSeconds(30));
                 logger.debug("Firing joinedCluster() event");
                 ClusterManager.fireJoinedCluster(true);
-            } else {
-                // Trigger event that a new node has joined the cluster
-                ClusterManager.fireJoinedCluster(nodeID.toByteArray(), true);
+
+                waitForClusterCacheToBeInstalled();
+
+                // Let the other nodes know that we joined the cluster
+                logger.debug("Done joining the cluster in split brain recovery. Now proceed informing other nodes that we joined the cluster.");
+                CacheFactory.doClusterTask(new NewClusterMemberJoinedTask(true));
             }
         }
         clusterNodesInfo.put(nodeID,
                 new HazelcastClusterNodeInfo(event.getMember(), cluster.getClusterTime()));
     }
 
+    /**
+     * Blocks the current thread until the cluster cache is guaranteed to support clustering. This is especially useful
+     * for executing cluster tasks immediately after joining. If this wait is not performed, the cache factory may still
+     * be using the 'default' strategy instead of the 'hazelcast' strategy, which leads to cluster tasks being silently
+     * discarded.
+     */
+    private void waitForClusterCacheToBeInstalled() {
```

Comment:
This `waitForClusterCacheToBeInstalled` solution sounds like a fix for a new problem: a problem that was already there, but from which we didn't suffer visibly, because we did not debug cluster tasks being fired as a direct result of an initital cluster join before. Please create a separate issue for this, so that we can document this problem. The fix seems OK to me. Doing a busy-loop isn't ideal, but if there's no callback/event that we can depend on... Does it make sense to add some kind of timeout that prevents a never-ending loop?

Reply:
Created new issue https://github.com/igniterealtime/openfire-hazelcast-plugin/issues/67.

Reply:
Introduced a 10 minute timeout in https://github.com/igniterealtime/openfire-hazelcast-plugin/pull/66/commits/83b1bbe399d37801b5f364c735e620b333782b0a.